src/core/crypto_tc.c

```diff
@@ -162,655 +105,1312 @@ int32_t Crypto_TC_ApplySecurity_Cam(const uint8_t* p_in_frame, const uint16_t in
+        status = CRYPTO_LIB_ERR_INVALID_SA_SERVICE_TYPE;
+    }
+    return status;
+}
 
-        // Try to assure SA is sane
-        status = crypto_tc_validate_sa(sa_ptr);
-        if (status != CRYPTO_LIB_SUCCESS)
+int32_t Crypto_TC_Handle_Enc_Padding(uint8_t sa_service_type, uint32_t* pkcs_padding, uint16_t* p_enc_frame_len, uint16_t* new_enc_frame_header_field_length, uint16_t tf_payload_len, SecurityAssociation_t* sa_ptr)
+{
+    int32_t status = CRYPTO_LIB_SUCCESS;
+    if (sa_service_type == SA_ENCRYPTION)
+    {
+        // Handle Padding, if necessary
+        if (sa_ptr->ecs == CRYPTO_CIPHER_AES256_CBC)
         {
-            mc_if->mc_log(status);
-            return status;
+            *pkcs_padding = tf_payload_len % TC_BLOCK_SIZE; // Block Sizes of 16
+
+            *pkcs_padding = TC_BLOCK_SIZE - *pkcs_padding; // Could potentially need 16 bytes of padding.
+
+            *p_enc_frame_len += *pkcs_padding; // Add the necessary padding to the frame_len + new pad length field
+
+            *new_enc_frame_header_field_length = (*p_enc_frame_len) - 1;
+#ifdef DEBUG
+
+            printf("SHPLF_LEN: %d\n", sa_ptr->shplf_len);
+            printf("Padding Needed: %d\n", *pkcs_padding);
+            printf("Previous data_len: %d\n", tf_payload_len);
+            printf("New data_len: %d\n", (tf_payload_len + *pkcs_padding));
+            printf("New enc_frame_len: %d\n", (*p_enc_frame_len));
+#endif
+            // Don't Exceed Max Frame Size! 1024
+            if (*p_enc_frame_len > TC_MAX_FRAME_SIZE)
+            {
+                status = CRYPTO_LIB_ERR_TC_FRAME_SIZE_EXCEEDS_SPEC_LIMIT;
+                mc_if->mc_log(status);
+            }
         }
+    }
+    return status;
+}
 
-#ifdef SA_DEBUG
-        printf(KYEL "DEBUG - Printing SA Entry for current frame.\n" RESET);
-        Crypto_saPrint(sa_ptr);
+int32_t Crypto_TC_Frame_Validation(uint16_t* p_enc_frame_len)
+{
+    int32_t status = CRYPTO_LIB_SUCCESS;
+    if (*p_enc_frame_len > current_managed_parameters->max_frame_size)
+    {
+#ifdef DEBUG
+        printf("Managed length is: %d\n", current_managed_parameters->max_frame_size);
+        printf("New enc frame length will be: %d\n", *p_enc_frame_len);
 #endif
+        printf(KRED "Error: New frame would violate maximum tc frame managed parameter! \n" RESET);
+        status = CRYPTO_LIB_ERR_TC_FRAME_SIZE_EXCEEDS_MANAGED_PARAM_MAX_LIMIT;
+        mc_if->mc_log(status);
+        printf("STATUS=%d\n", status);
+        return status;
+    }
+    // Ensure the frame to be created will not violate spec max length
+    if ((*p_enc_frame_len > 1024) && status == CRYPTO_LIB_SUCCESS)
+    {
+        printf(KRED "Error: New frame would violate specification max TC frame size! \n" RESET);
+        status = CRYPTO_LIB_ERR_TC_FRAME_SIZE_EXCEEDS_SPEC_LIMIT;
+        mc_if->mc_log(status);
+    }
+    return status;
+}
+
+int32_t Crypto_TC_Accio_Buffer(uint8_t** p_new_enc_frame, uint16_t* p_enc_frame_len)
```

Comment:
Is this function necessary? Should we just have this memory pre-allocated in the library and cleared for use each call?

Reply:
This is necessary for TC Encrypt on the ground due to KMC.